tsl/src/compression/compression.c

```diff
@@ -2042,3 +2047,226 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
+
+				/*
+				 * In PG versions <= 14 NULL values are always considered distinct
+				 * from other NULL values and therefore NULLABLE multi-columnn
+				 * unique constraints might expose unexpected behaviour in the
+				 * presence of NULL values.
+				 * Since SK_SEARCHNULL is not supported by heap scans we cannot
+				 * build a ScanKey for NOT NULL and instead have to do those
+				 * checks manually.
+				 */
+				if (isnull)
+				{
+					*null_columns = bms_add_member(*null_columns, attno);
+				}
+				else
+				{
+					ScanKeyEntryInitialize(&scankeys[index],
+										   0, /* flags */
+										   attno,
+										   BTEqualStrategyNumber,
+										   InvalidOid, /* No strategy subtype. */
+										   decompressor.in_desc->attrs[attno - 1].attcollation,
+										   tce->eq_opr_finfo.fn_oid,
+										   value);
+					index++;
+				}
+			}
+		}
+	}
+
+	*num_scankeys = index;
+	return scankeys;
+}
+
+void
+decompress_batches_for_insert(Chunk *chunk, TupleTableSlot *slot)
+{
+	Relation out_rel = table_open(chunk->table_id, RowExclusiveLock);
+
+	if (!ts_indexing_relation_has_primary_or_unique_index(out_rel))
+	{
+		/*
+		 * If there are no unique constraints there is nothing to do here.
+		 */
+		table_close(out_rel, RowExclusiveLock);
+		return;
+	}
+
+	Chunk *comp = ts_chunk_get_by_id(chunk->fd.compressed_chunk_id, true);
+	Relation in_rel = relation_open(comp->table_id, RowExclusiveLock);
+
+	RowDecompressor decompressor = build_decompressor(in_rel, out_rel);
+	Bitmapset *unique_columns = relation_unique_columns(out_rel);
+	Bitmapset *null_columns = NULL;
+
+	int num_scankeys;
+	ScanKeyData *scankeys = build_scankeys(chunk->fd.hypertable_id,
+										   decompressor,
+										   unique_columns,
+										   &null_columns,
+										   slot,
+										   &num_scankeys);
+
+	TableScanDesc heapScan =
+		table_beginscan(in_rel, GetTransactionSnapshot(), num_scankeys, scankeys);
+
+	for (HeapTuple compressed_tuple = heap_getnext(heapScan, ForwardScanDirection);
+		 compressed_tuple != NULL;
+		 compressed_tuple = heap_getnext(heapScan, ForwardScanDirection))
+	{
+		Assert(HeapTupleIsValid(compressed_tuple));
+		bool valid = true;
+
+		/*
+		 * Since the heap scan API does not support SK_SEARCHNULL we have to check
+		 * for NULL values manually when those are part of the constraints.
+		 */
+		for (int attno = bms_next_member(null_columns, -1); attno >= 0;
+			 attno = bms_next_member(null_columns, attno))
+		{
+			if (!heap_attisnull(compressed_tuple, attno, decompressor.in_desc))
+			{
+				valid = false;
+				break;
+			}
+		}
+
+		/*
+		 * Skip if NULL check failed.
+		 */
+		if (!valid)
+			continue;
+
+		heap_deform_tuple(compressed_tuple,
+						  decompressor.in_desc,
+						  decompressor.compressed_datums,
+						  decompressor.compressed_is_nulls);
+
+		row_decompressor_decompress_row(&decompressor);
+
+		TM_FailureData tmfd;
+		table_tuple_delete(in_rel,
```

Comment:
Suggest handling return value here, or at least asserting the result if only one possibility.

Reply:
done